File: src/convert.py
import re
from typing import *
from curve import *
def pathToCharacterWithCurve(path: str) -> List[Curve]:
    def stringToNum(s: str) -> Num:
        """字符转数字"""
        if '.' in s:
            return float(s)
        else:
            return int(s)

    def paramsStrToParamsList(paramsStr: str) -> List[Num]:
        paramsList = []
        for item in re.split('[,\s]+|(?<=\d)(?=-)', paramsStr.strip()):
            paramsList.append(stringToNum(item))
        return paramsList
    headPoint:Optional[Point] = None
    nowPoint:Point = Point(0, 0)
    characterWithCurve:List[Curve] = []
    first = re.split('(?<=\d)\s*(?=[A-Za-z])', path.strip())
    for item in first:
        typeName, paramsStr = re.split('(?<=[A-Za-z])', item)
        if paramsStr:
            paramsList = paramsStrToParamsList(paramsStr)
        if typeName.lower() == 'm':
            end = Point(paramsList[0], paramsList[1])
            if typeName.islower():
                end.offset(nowPoint.x, nowPoint.y)
            headPoint = end
            nowPoint = end
        elif typeName.lower() == 'l':
            end = Point(paramsList[0], paramsList[1])
            if typeName.islower():
                end.offset(nowPoint.x, nowPoint.y)
            if end.isSamePosition(headPoint):
                end = headPoint
                headPoint = None
            characterWithCurve.append(CurveL(nowPoint, end))
            nowPoint = end
        elif typeName.lower() == 'h':
            end = Point(paramsList[0], nowPoint.y)
            if typeName.islower():
                end.offset(nowPoint.x, 0)
            if end.isSamePosition(headPoint):
                end = headPoint
                headPoint = None
            characterWithCurve.append(CurveL(nowPoint, end))
            nowPoint = end
        elif typeName.lower() == 'v':
            end = Point(nowPoint.x, paramsList[0])
            if typeName.islower():
                end.offset(0, nowPoint.y)
            if end.isSamePosition(headPoint):
                end = headPoint
                headPoint = None
            characterWithCurve.append(CurveL(nowPoint, end))
            nowPoint = end
        elif typeName.lower() == 'q':
            end = Point(paramsList[0], paramsList[1])
            control = Point(paramsList[2], paramsList[3])
            if typeName.islower():
                end.offset(nowPoint.x, nowPoint.y)
                control.offset(nowPoint.x, nowPoint.y)
            if end.isSamePosition(headPoint):
                end = headPoint
                headPoint = None
            characterWithCurve.append(CurveQ(nowPoint, end, control))
            nowPoint = end
        elif typeName.lower() == 't':
            pass
        elif typeName.lower() == 'c':
            pass
        elif typeName.lower() == 's':
            pass
        elif typeName.lower() == 'a':
            pass
        elif typeName.lower() == 'z':
            if headPoint:
                if not headPoint.isSamePosition(nowPoint):
                    characterWithCurve.append(CurveL(nowPoint, headPoint))
                nowPoint = headPoint
            pass
    return characterWithCurve

File: src/svg/convert.ts
import { Point, Curve, CurveL, CurveQ, Coordinate } from './class'
//start
export function pathToCurveList(path: string): Array<Curve> {
  let startPoint: Point = new Point(0, 0)
  let lastPoint: Point = startPoint
  let curves: Array<Curve> = []
  let operatorStrs = path.trim().split(/(?<=\d|[Zz])\s*(?=[A-Za-z])/)
  for (let operatorStr of operatorStrs) {
    let params: number[] = []
    let [typeName, paramsStr] = operatorStr.split(/(?<=[A-Za-z])/)
    if (paramsStr) {
      params = paramsStr
        .trim()
        .split(/[,\s]+|(?<=\d)(?=-)/)
        .map((value) => Number(value))
    }
    if (typeName.toUpperCase() == 'M') {
      if (typeName == 'm') {
        startPoint = new Point(lastPoint.x + params[0], lastPoint.y + params[1])
        lastPoint = startPoint
      } else {
        startPoint = new Point(params[0], params[1])
        lastPoint = startPoint
      }
    } else if (typeName.toUpperCase() == 'L') {
      let endCoordinate = { x: params[0], y: params[1] }
      if (typeName == 'l') {
        endCoordinate.x += lastPoint.x
        endCoordinate.y += lastPoint.y
      }
      let end = startPoint.isSameCoordinate(endCoordinate)
        ? startPoint
        : new Point(endCoordinate.x, endCoordinate.y)
      curves.push(new CurveL(lastPoint, end))
      lastPoint = end
    } else if (typeName.toUpperCase() == 'Q') {
      let control = new Point(params[0], params[1])
      let endCoordinate = { x: params[2], y: params[3] }
      if (typeName == 'q') {
        control.offset(lastPoint.x, lastPoint.y)
        endCoordinate.x += lastPoint.x
        endCoordinate.y += lastPoint.y
      }
      let end = startPoint.isSameCoordinate(endCoordinate)
        ? startPoint
        : new Point(endCoordinate.x, endCoordinate.y)
      curves.push(new CurveQ(lastPoint, end, control))
      lastPoint = end
    }
    //H V T S C A Z暂时用不上，省略了
  }
  return curves
}
